sql/vector_mhnsw.cc

```diff
@@ -188,7 +188,56 @@ struct FVector
   }
 #endif
 
+
+/*
+  ARM NEON implementation. A microbenchmark shows 1.7x dot_product() performance
+  improvement compared to regular -O2/-O3 builds and 2.4x compared to builds
+  with auto-vectorization disabled.
+
+  There seem to be no performance difference between vmull+vmull_high and
+  vmull+vmlal2_high implementations. The second implementation is preserved
+  just for reference.
+*/
+
+#ifdef NEON_IMPLEMENTATION
+  static constexpr size_t NEON_bytes= 128 / 8;
+  static constexpr size_t NEON_dims= NEON_bytes / sizeof(int16_t);
+
+  static float dot_product(const int16_t *v1, const int16_t *v2, size_t len)
+  {
+    int64_t d= 0;
+    for (size_t i= 0; i < (len + NEON_dims - 1) / NEON_dims; i++)
+    {
+      int16x8_t p1= vld1q_s16(v1);
+      int16x8_t p2= vld1q_s16(v2);
+#if 0
+      d+= vaddlvq_s32(vmlal_high_s16(vmull_s16(vget_low_s16(p1),
+                                               vget_low_s16(p2)), p1, p2));
+#else
+      d+= vaddlvq_s32(vmull_s16(vget_low_s16(p1), vget_low_s16(p2))) +
+          vaddlvq_s32(vmull_high_s16(p1, p2));
+#endif
+      v1+= NEON_dims;
+      v2+= NEON_dims;
+    }
```

Comment:
did you look at the assembly code for this function? Can you paste it here, please? (not in the code, in the comment on a PR).

Reply:
another question, would be good to compare with the code that compiler would generate without explicit use of intrinsics, but with explicit vectorization. That is, with something like
```c
typedef float v128i __attribute__((vector_size(NEONbytes)));
v128i *p1= (v128i*)v1, *p2=(v128i*)v2, d=0;
for (size_t i= 0; i < (len + NEON_dims - 1) / NEON_dims; i++, p1++, p2++)
  d+=*p1 * *p2;
```
This compiles into (https://godbolt.org/z/GjnsqoEhq)
```
dot_product(void const*, void const*, int):
        add     w4, w2, 14
        adds    w3, w2, 7
        csel    w4, w4, w3, mi
        asr     w4, w4, 3
        cmp     w2, 0
        ble     .L4
        movi    v0.4s, 0
        mov     x2, 0
        mov     w3, 0
.L3:
        ldr     q2, [x0, x2]
        add     w3, w3, 1
        ldr     q1, [x1, x2]
        add     x2, x2, 16
        mla     v0.4s, v2.4s, v1.4s
        cmp     w3, w4
        blt     .L3
        dup     s1, v0.s[1]
        add     v0.2s, v1.2s, v0.2s
        fmov    w0, s0
        ret
.L4:
        mov     w0, 0
        ret
```
and your explicit code is
```
dot_product(short const*, short const*, int):
        add     w5, w2, 14
        adds    w3, w2, 7
        csel    w5, w5, w3, mi
        mov     x6, x0
        asr     w5, w5, 3
        cmp     w2, 0
        ble     .L4
        mov     x3, 0
        mov     w4, 0
        mov     w0, 0
.L3:
        ldr     q0, [x6, x3]
        add     w4, w4, 1
        ldr     q2, [x1, x3]
        add     x3, x3, 16
        smull   v1.4s, v0.4h, v2.4h
        smull2  v0.4s, v0.8h, v2.8h
        saddlv  d1, v1.4s
        saddlv  d0, v0.4s
        add     d0, d1, d0
        fmov    x2, d0
        add     w0, w0, w2
        sxth    w0, w0
        cmp     w4, w5
        blt     .L3
        ret
.L4:
        mov     w0, 0
        ret
```
the `#ifdef 0` implementation differs in
```
        smull   v0.4s, v2.4h, v1.4h
        smlal2  v0.4s, v2.8h, v1.8h
        saddlv  d0, v0.4s
        fmov    x2, d0
```

Reply:
Dunno, I tried adding `dot_product()` w/o explicit vectorization and adding `-ftree-vectorize`. It produced something very complex that had vector instructions. Probably it couldn't find a way to get your example vectorized due to widening or whatnot.